### DISCUSSION TITLE
New next.js instance regenerates shared cache

### DISCUSSION THREAD

**Discussion by TimGeudens** [vercel/next.js#67842] [Help]
### Summary

We are self-hosting a next.js app using app router. We have a shared disk where all instances run from. Since we use the default next.js cache it is also used for reading and writing the fetch cache and full route cache. 

However we noticed that the full route cache files (.html, .meta & .rsc) are constantly regenerated. This happens every 3-10 minutes. 
After doing some testing locally we noticed that a new server instance will regenerate those files no matter what. 
On production when an instance restarts or when it is added by automatic scaling, this triggers a regeneration of full route files for all requests on that new instance. As a result we get quite some disk writes which we did not expect. The fetch cache however seems fine since rebuilding the page is still pretty fast. 

We also tried adding a Redis cache handler but saw the same issue there (cache string gets updated when new instance comes online). 

We also tried adding the build id in the config when testing locally but that did not help either. 
`  generateBuildId: async () => {
    return 'build-1';
  },
  deploymentId: 'build-1',`
  
Is this something that is expected or is there something we could change so that new instances just read from the existing cache? 

### Additional information

```js
Some more info about the app: we have one catch-all route (next to some fixes routes) that generates 98% of the pages. 
We added this to enable ISR:
`export const dynamicParams = true;
export const generateStaticParams = async () => [];`

We disabled memory cache in config with
`  cacheMaxMemorySize: 0
`

Then we just ran multiple instances like this:
`    "start": "next start -p 3003",
    "start2": "next start -p 3004",
    "start3": "next start -p 3005",`

You can check the timestamp of the generated .html file in .next/server/app/your-page.html and you can see the date changes when restarting an instance and visiting /your-page.
```

**Comment by wgbx** [comment]
I have the same problem, is there a better plan? Thank you

**Comment by mchlbld** [comment]
We have a custom cache handler using Redis and encountered the same issue.
The [documentation](https://dev.azure.com/draeger/DraegerWeb/_git/Draeger.Web.Frontend.NextJS/pullrequest/114728?_a=files) states that the FRC is cleared on **deployment** not on **restart** which is something different from my perspective.

Looking [in the code](https://github.com/vercel/next.js/blob/504fa32ae6276f15e2662f52f9ba8203362f227c/packages/next/src/server/lib/incremental-cache/index.ts#L516) I found out, that cache entries are considered stale when the `lastModified` field is older than `performance.timeOrigin` which makes no sense to me as `performance.timeOrigin` does not necessarily mark a deployment but a simple restart.

Setting `lastModified` to 0 fixes the issue but anyway the general behavior is quite confusing.

**Comment by leejpsd** [comment]
This is reproducible from the source, and your instinct is right — it's per-instance state, but not the one you patched with `generateBuildId`.

The revalidate period for a route is resolved through `SharedCacheControls` (`server/lib/incremental-cache/shared-cache-controls.external.js`), which checks two places: a **process-local in-memory `Map`**, populated only when *this* instance writes an entry, and the **prerender manifest**, which only contains build-time prerendered paths. Your setup generates ~98% of pages at runtime (`generateStaticParams: async () => []`), so those routes are in neither place on a freshly booted instance.

When that lookup comes back empty, `IncrementalCache` falls back to a default (`server/lib/incremental-cache/index.js`):

```js
const initialRevalidateSeconds = cacheControl ? cacheControl.revalidate : isFallback ? false : 1;
```

**One second.** So the first request for each route on a new instance finds the shared entry, treats it as stale (`revalidateAfter = lastModified + 1s`, long past), serves it stale-while-revalidate, re-renders in the background, and rewrites the files. The `set()` then populates that instance's in-memory map with the real revalidate, so it settles — until the next instance boots and repeats it. With autoscaling churn, that's your constant disk writes.

This also explains the two things you observed:

- **The Redis handler didn't help** (@mchlbld saw the same): the stale decision happens *above* the cache handler, in `IncrementalCache`, using the handler's `lastModified` plus this resolved revalidate. A custom handler can't override it.
- **`generateBuildId` / `deploymentId` didn't help**: neither participates in this lookup at all (`deploymentId` never reaches the incremental cache).

What you can actually do about it: not much from userland, honestly. Returning real paths from `generateStaticParams` fixes it only for those specific paths (they land in the manifest). One reassurance: the user never waits on this — the stale copy is served immediately; the cost is the background render and the writes, not latency.

The design-level fix is upstream: the entry itself carries no revalidate, so a fresh process can't know it. Notably the newer plural `cacheHandlers` interface (for `'use cache'`) already fixed this shape — entries there carry `revalidate`/`expire` inline. The singular handler's entries could do the same. Given this discussion has been open since 2024 with several confirmations, it may be worth distilling into an issue — the analysis above should transfer directly.

(Context for where this comes from: I maintain a Redis cache handler for self-hosted setups and hit this same wall — it's the one thing that can't be fixed from the handler layer.)
